libslide/src/evaluator_rules/pattern_matching.rs

```diff
@@ -122,78 +152,109 @@ impl Replacements {
         Some(replacements)
     }
 
-    fn insert(&mut self, k: ExprPat, v: Expr) -> Option<Expr> {
-        self.map.insert(k, v)
+    fn insert(&mut self, k: &Rc<ExprPat>, v: Rc<Expr>) -> Option<Rc<Expr>> {
+        self.map.insert(hash(k.as_ref()), v)
     }
```

Comment:
isn't k already a reference from its definition in line 155. why do we need it again as_ref()

Reply:
It's a reference to the `Rc`, not the inner data. We call `as_ref` to be a reference (pointer) to the underlying data, which is the common subexpression.